controllers/openstackfloatingippool_controller.go

```diff
@@ -0,0 +1,366 @@
+			return ctrl.Result{Requeue: true}, r.Client.Update(context.Background(), pool)
+		}
+	} else {
+		// Handle deletion
+		return r.reconcileDelete(ctx, pool)
+	}
+
+	scope, err := r.ScopeFactory.NewClientScopeFromFloatingIPPool(ctx, r.Client, pool, r.CaCertificates, log)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if err := r.setIPStatuses(ctx, scope, pool); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	patchHelper, err := patch.NewHelper(pool, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		if err := patchHelper.Patch(ctx, pool); err != nil {
+			if reterr == nil {
+				reterr = fmt.Errorf("error patching OpenStackFloatingIPPool %s/%s: %w", pool.Namespace, pool.Name, err)
+			}
+		}
+	}()
+
+	claims := &ipamv1.IPAddressClaimList{}
+	if err := r.Client.List(context.Background(), claims, client.InNamespace(req.Namespace), client.MatchingFields{infrav1.OpenStackFloatingIPPoolNameIndex: pool.Name}); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	for _, claim := range claims.Items {
+		claim := claim
+		log := log.WithValues("claim", claim.Name)
+
+		cluster, err := util.GetClusterFromMetadata(ctx, r.Client, claim.ObjectMeta)
+		if err != nil {
+			log.Info("Could not get cluster resource for IPAddressClaim")
+			return ctrl.Result{}, nil
+		}
+
+		infraCluster, err := r.getInfraCluster(ctx, cluster, &claim)
+		if err != nil {
+			return ctrl.Result{}, errors.New("error getting infra provider cluster")
+		} else if infraCluster == nil {
+			log.Info("infra cluster is not ready yet")
+			return ctrl.Result{}, nil
+		}
+
+		if claim.Status.AddressRef.Name == "" {
+			clusterName := fmt.Sprintf("%s-%s", claim.ObjectMeta.Labels[clusterv1.ClusterNameLabel], claim.Namespace)
+			ip, err := r.getIP(ctx, scope, pool, infraCluster, clusterName)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+
+			ipAddress := &ipamv1.IPAddress{
+				ObjectMeta: ctrl.ObjectMeta{
+					Name:      claim.Name,
+					Namespace: claim.Namespace,
+					OwnerReferences: []metav1.OwnerReference{
+						{
+							APIVersion: claim.APIVersion,
+							Kind:       claim.Kind,
+							Name:       claim.Name,
+							UID:        claim.UID,
+						},
+					},
+				},
+				Spec: ipamv1.IPAddressSpec{
+					ClaimRef: corev1.LocalObjectReference{
+						Name: claim.Name,
+					},
+					PoolRef: corev1.TypedLocalObjectReference{
+						APIGroup: pointer.String(infrav1.GroupVersion.Group),
+						Kind:     pool.Kind,
+						Name:     pool.Name,
+					},
+					Address: ip,
+				},
+			}
+
+			if !contains(pool.Spec.PreAllocatedFloatingIPs, ip) && pool.Spec.ReclaimPolicy == infrav1.ReclaimDelete {
+				controllerutil.AddFinalizer(ipAddress, infrav1.DeleteFloatingIPFinalizer)
+			}
+
+			if err = r.Client.Create(ctx, ipAddress); err != nil {
+				return ctrl.Result{}, err
+			}
+
+			claim.Status.AddressRef.Name = ipAddress.Name
+			if err = r.Client.Status().Update(ctx, &claim); err != nil {
```

Comment:
It's ok to update the claim here, too.

Reply:
I’ll have a look at this soon, I realized when looking at the other pr for the is machine controller, it is watching and would need the entire claim to be updated. I’ll need to have a look at the ipam proposal again though, if it mentions if the ipamprovider only needs to update claim.Status I think I should handle it in the machine controller instead so we support any ipamprovider that follows the spec

Reply:
My previous comment may not have been clear: I think this update is good and you don't need to change anything.